app/src/org/commcare/utils/DeviceIdentifier.java

```diff
@@ -0,0 +1,41 @@
+package org.commcare.utils;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import androidx.preference.PreferenceManager;
+import java.util.UUID;
+
+/**
+ * A Utility to get a unique identifier for app's installation.
+ * @author $|-|!˅@M
+ */
+public final class DeviceIdentifier {
+
+    private static volatile String uuid;
+    private static final String PREFIX = "commcare_";
+    private static final String KEY_DEVICE_IDENTIFIER = "commcare_device_id";
+
+    /**
+     * Returns a unique(very highly likely) identifier for this device.
+     * This identifier is prefixed with "commcare_"
+     *
+     * This id may change when user re-installs the app or does a factory reset.
+     */
+    public static String getDeviceIdentifier(Context context) {
+        if (uuid == null) {
+            synchronized (DeviceIdentifier.class) {
+                if (uuid == null) {
+                    SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
```

Comment:
Using SharedPreference to store this unique id. We can also choose to create a text file and store UUID to that file. I'm not sure which option is more suitable of the two so using the sharedpreferences approach here.